core/utils/fifo_opener.cc

```diff
@@ -0,0 +1,304 @@
+    close(fd);
+    return -error;
+  }
+
+  // Store new descriptor and update generation.  There's nothing
+  // we're competing with (no new thread could be started and
+  // MarkDead does nothing with old_fd being kNotOpen), so a
+  // simple store suffices.
+  wrapped_fd_.store(Wrap(fd, gen + 1));
+  return 0;
+}
+
+bool FifoOpener::OpenInThread() {
+  std::lock_guard<std::mutex> guard(threadlock_);
+  if (shutting_down_) {
+    return false;
+  }
+  if (!opening_) {
+    StartThreadLocked();  // this sets opening_ to true
+  }
+  return true;
+}
+
+/*
+ * Internal: start the opener thread.  Caller holds the lock.
+ */
+void FifoOpener::StartThreadLocked() {
+  assert(!opening_);
+  if (thread_.Start()) {
+    opening_ = true;
+  } else {
+    LOG_FIRST_N(ERROR, 1) << "error starting FIFO opener thread: "
+                          << std::strerror(errno);
+  }
+}
+
+void FifoOpenerThread::Run() {
+  int fd;
+
+  do {
+    // Do a blocking open().
+    fd = open(fifopath_, O_WRONLY);
+  } while (fd < 0 && !IsExitRequested());
+
+  // It's open, so we're almost done.  Defer exit requests so
+  // we can make system calls without getting interrupted by
+  // a knock thread if there is one (or one starts here).
+  PushDefer();
+
+  // Set up the file descriptor: it should be non-blocking.
+  // It also needs additional setup, which is something our
+  // owner/instantiator will provide.
+  if (fd < 0) {
+    fd = FifoOpener::kNotOpen;
+  } else if (!SetNonblocking(fd) || !owner_->InitFifo(fd)) {
+    close(fd);
+    fd = FifoOpener::kNotOpen;
+  }
+
+  // We could test for exit-requested here, but there's no real
+  // point - we can just hand over the descriptor.  Anyone poking
+  // us to terminate has to wait for us to exit since we might
+  // race anyway, so just give it over to our owner.
+  owner_->SetNewFdFromThread(fd);
+}
+
+FifoOpener::~FifoOpener() {
+  // Stop reconnects.  (Shouldn't be required, but feels safer.)
+  reconnect_ = false;
+
+  // If there is an opener thread, stop it now and wait for it to
+  // finish.  In any case this is the end of this FifoOpener instance,
+  // so with reconnect_ cleared, there will be no more opener threads.
+  Shutdown();
+}
+
+/*
+ * Stops any opener thread.  Once done, everything is back to the
+ * way it was before an opener thread was started (or before
+ * opening the FIFO directly).  If there was an open fd, it is
+ * closed now.
+ */
+void FifoOpener::Shutdown() {
+  {
+    std::lock_guard<std::mutex> guard(threadlock_);
+    shutting_down_ = true;
+  }
+  thread_.Terminate(bess::utils::SyscallThreadAny::Wait);
+  {
+    std::lock_guard<std::mutex> guard(threadlock_);
+
+    // While holding the lock (preventing other opens),
+    // mark "not open" and close any open FIFO.  Reset
+    // the thread so that new opens can happen if desired.
+    // Then we're done with this shutdown.
+    SetNewFdLocked(kNotOpen);
+    thread_.Reset();
+    shutting_down_ = false;
+  }
+}
+
```

Comment:
May I don't understand the code Push/PopDefer() clearly. Should the `PopDefer()` be called when shutdown?

Reply:
It's not required, the push/pop is just in case there's some future need for this.  For instance, if you had a thread where you needed to make some system call (`socket`, `open`, whatever) that was blocking but should get interrupted, but then if that call _works_ you also need to make several more system calls that must not be interrupted, then go back into the interruptible loop, you could `PushDefer(); syscall1(arg); syscall2(arg); PopDefer(); continue;` for instance.

(In other words, right now this is just over-engineered 😈 but since they're inline it should not be adding any actual runtime overhead)